bpo-40066: Enum: modify `repr()` and `str()`

ethanfurman opened the pull request:
Enumerations created from `Enum._convert_` will have their `repr()` changed to be

  `module.member_name`

and their `str()` to be

  `member_name`

The exception is `StrEnum`, whose `str()` will be

  `value`

as in, the member's value.

<!-- issue-number: [bpo-40066](https://bugs.python.org/issue40066) -->
https://bugs.python.org/issue40066
<!-- /issue-number -->

ethanfurman committed:
add new repr and str for converted Int-Enums/Flags

ethanfurman committed:
wip

ethanfurman committed:
make Enum.__str__ print name or value

non-StrEnum `__str__` will print the member name
StrEnum `__str__` will print the member value

ethanfurman committed:
add news entry

ethanfurman committed:
fix tests that checked Enum str()

ethanfurman committed:
wip

create enum._stdlib_enum decorator for Enums created the normal way: it
adjusts `__repr__` and `__str__` to be the same as if the Enum had been
created using `_convert_`

update various modules to use the new decorator

ethanfurman committed:
fix format for global string override

if `enum.__str__` is either `Enum.__str__` or `global_enum_str` use the
value's `__format__`

ethanfurman committed:
update docs with new stdlib enum reprs and strs

ethanfurman committed:
plain Enum has updated str() and repr()

repr() = class.member_name
str() = member_name

ethanfurman committed:
_stdlib_enum -> global_enum

rename to `global_enum` and have `global_enum` update the module's
namespace with the members

remove `_stdlib_enum` where the enums were not being exported to the
module's namespace

ethanfurman committed:
fix pydoc and unicode tests

ethanfurman committed:
fix white space

ethanfurman committed:
fix signal tests

ethanfurman committed:
update NEWS entry

gvanrossum commented on Lib/enum.py:
```diff
@@ -946,3 +942,42 @@ def _decompose(flag, value):
+        cls.__repr__ = global_flag_repr
+    else:
+        cls.__repr__ = global_int_repr
+    vars(sys.modules[cls.__module__]).update(cls.__members__)
```
Does using `vars(x)` instead of `x.__dict__` buy you anything here?

ethanfurman replied:
No, it's just what I'm used to using to get a class' namespace while I'm working on it.  Inside the class body I could use `locals()`, but then I have to remember if `locals()` is well-defined for classes, and `vars()` works whether I'm in the class body or in the metaclass.

I can change it if you think `.__dict__` is more appropriate.

gvanrossum replied:
Please do. `vars()` is an end user feature and mostly meant for the REPL. Here we're just updating a module's dict.

gvanrossum commented on Lib/re.py:
```diff
@@ -154,31 +155,6 @@ class RegexFlag(enum.IntFlag):
-    __str__ = object.__str__
-
-globals().update(RegexFlag.__members__)
-
```
Nice win to delete/move all this code!

ethanfurman replied:
Thanks.  It took me a few iterations to figure out how I wanted it, and some insights from Serhiy helped a great deal.

gvanrossum commented on Misc/NEWS.d/next/Library/2020-09-23-21-58-34.bpo-40066.f1dr_5.rst:
```diff
@@ -0,0 +1,4 @@
+Enum's `repr()` and `str()` have changed: `repr()` is now *EnumClass.MemberName*
+and `str()` is *MemberName*.  Additionally, stdlib Enum's whose contents are
+available as module attributes, such as `RegexFlag.IGNORECASE`, have their
+`repr()` as *module.name*, e.g. `re.IGNORECASE`.
```
Maybe also add a section to what's new in 3.10?

ethanfurman replied:
I thought `blurb` did that?

gvanrossum replied:
No, blurb just updates Misc/NEWS, which is an endless list of things (though more focused than the commit log :-).

What's new in 3.10 is a separate doc written by human editor who curate the most important news and group it by affected area.

gvanrossum replied:
https://docs.python.org/3.10/whatsnew/3.10.html

serhiy-storchaka replied:
Also the `versionchanged` directive is needed in the module documentation.

gvanrossum reviewed (commented):
Nice. How did you track down all the docs that need to be changed? (I note there aren't any doc changes do to the change in str().)

serhiy-storchaka commented on Doc/library/enum.rst:
```diff
@@ -667,28 +667,28 @@ It is also possible to name the combinations::
+    Perm.RWX
     >>> ~Perm.RWX
-    <Perm.-8: -8>
+    Perm.-8
```
In this case I would prefer `~Perm.RWX` or `Perm(-8)`. But this may be a different issue.

ethanfurman replied:
Negative values are no longer a thing -- they are converted to their positive equivalent.

serhiy-storchaka commented on Doc/library/enum.rst:
```diff
@@ -667,28 +667,28 @@ It is also possible to name the combinations::
 
     >>> Perm.R & Perm.X
-    <Perm.0: 0>
+    Perm.0
```
`Perm(0)` would look better.

ethanfurman replied:
I went with `0x0` to match other non-Flag values, similarly to the way `re.RegexFlag` does it.

serhiy-storchaka commented on Doc/library/enum.rst:
```diff
@@ -667,28 +667,28 @@ It is also possible to name the combinations::
 
     >>> Perm.X | 8
-    <Perm.8|X: 9>
+    Perm.8|Perm.X
```
And here I would prefer `Perm.X|8`.

ethanfurman replied:
It is now `Perm.X|0x8`

serhiy-storchaka commented on Doc/library/enum.rst:
```diff
@@ -1198,7 +1198,7 @@ constructor.  For example::
     ...     example = '11', 16      # '11' will be interpreted as a hexadecimal
     ...                             # number
     >>> MyEnum.example
```
I think that it would more useful to use `MyEnum.example.value` or `int(MyEnum.example)` in the example.

ethanfurman replied:
Yes, I added the `.value` here and in other places.

serhiy-storchaka commented on Lib/enum.py:
```diff
@@ -76,8 +77,10 @@ def __setitem__(self, key, value):
-                raise ValueError(f'_sunder_ names, such as "{key}", are '
-                                  'reserved for future Enum use')
+                raise ValueError(
+                        '_sunder_ names, such as %r, are reserved for future Enum use'
```
You can use f-string with `!r`.

ethanfurman replied:
Yes, but I'm not using f-strings anywhere else.

ethanfurman replied:
At some point I'll switch all the %-interpolations to use f-strings.

serhiy-storchaka commented on Lib/enum.py:
```diff
@@ -946,3 +942,42 @@ def _decompose(flag, value):
+    for m in self.__class__:
+        if value & m._value_:
+            value &= ~m._value_
+            members.append('re.%s' % (m._name_, ))
```
`re`?

ethanfurman replied:
Oops.  Looks like I need to add a test for that branch of code.

serhiy-storchaka commented on Lib/enum.py:
```diff
@@ -946,3 +942,42 @@ def _decompose(flag, value):
     return members, not_covered
+
+def global_int_repr(self):
+    return '%s.%s' % (self.__class__.__module__, self.name)
```
What if add an attribute to the enum class which is set to the class name by default and can be set to the module name for enums whose members are exposed as globals?
```python
    return '%s.%s' % (self.__class__._namespacename, self.name)
```
It will make easier customization of the repr.

ethanfurman replied:
I think writing a custom `__repr__` is not that difficult.

serhiy-storchaka commented on Lib/enum.py:
```diff
@@ -946,3 +942,42 @@ def _decompose(flag, value):
+            value &= ~m._value_
+            members.append('re.%s' % (m._name_, ))
+    if value:
+        members.append(hex(value))
```
For general code I would add a class method for converting value to string. Depending on the class octal or binary may be better than hexadecimal.

ethanfurman replied:
Fair point.

ethanfurman committed:
use `.__dict__` instead of `vars()`

iritkatriel commented on Lib/enum.py:
```diff
@@ -789,27 +791,23 @@ def __iter__(self):
+            self._name_ = '%s' % '|'.join([
+                    str(m._name_ or m._value_) for m in members
+                    ])
+        return self._name_
```
Now whichever of `__str__` and `__repr__` is called first will initialize `_name_` for all time. Even if they do the same thing now, they might drift in the future.  Should this initialization code be in a common function to make sure it's consistent?    One option would be to override name() and do it there.

ethanfurman replied:
The name is now initialized when the pseudo-member is created.

ethanfurman committed:
wip

ethanfurman committed:
minor updates

* rename global_int_repr --> global_enum_repr
* adjust global_flag_repr to print hex number for out-of-range values
* extract function when _generate_next_value_ is a static method

ethanfurman committed:
rewrite enum docs; add enum how-to

ethanfurman committed:
Enum: streamline repr() and str(); improve docs

- repr() is now ``enum_class.member_name``
- str() is now ``member_name``
- add HOW-TO section for ``Enum``
- change main documentation to be an API reference

ethanfurman committed:
add ``whatsnew`` entry

ethanfurman committed:
Update 3.10.rst
